ec2/sc-ec2-linux-docker-notebook.yaml

```diff
@@ -0,0 +1,513 @@
+          Effect: 'Deny'
+          Principal: "*"
+          Condition:
+            StringNotEquals:
+              "aws:PrincipalArn": !Sub "${InstanceRole.Arn}"
+          Action:
+            - kms:Decrypt
+          Resource: '*'
+        - Sid: Allow use of the key
+          Effect: Allow
+          Principal:
+            AWS: !Sub "${InstanceRole.Arn}"
+          Action:
+          - kms:DescribeKey
+          - kms:Encrypt
+          - kms:Decrypt
+          - kms:ReEncrypt*
+          - kms:GenerateDataKey
+          - kms:GenerateDataKeyWithoutPlaintext
+          Resource: '*'
+        - Sid: 'Default KMS Policy that enables IAM permissions'  # https://docs.aws.amazon.com/kms/latest/developerguide/key-policies.html#key-policy-default
+          Effect: 'Allow'
+          Principal:
+            AWS: !Sub 'arn:aws:iam::${AWS::AccountId}:root'
+          Action:
+            - 'kms:*'
+          Resource: '*'
+        - Sid: "Allow administration of the key to CFN service role"
+          Effect: "Allow"
+          Principal:
+            AWS:
+              - !ImportValue
+                'Fn::Sub': '${AWS::Region}-sc-ec2vpc-launchrole-LaunchRoleArn'
+          Action:
+            - "kms:Create*"
+            - "kms:Describe*"
+            - "kms:Enable*"
+            - "kms:List*"
+            - "kms:Put*"
+            - "kms:Update*"
+            - "kms:Revoke*"
+            - "kms:Disable*"
+            - "kms:Get*"
+            - "kms:Delete*"
+            - "kms:ScheduleKeyDeletion"
+            - "kms:CancelKeyDeletion"
+          Resource: "*"
+
+  SSMParameterEncryptionKeyAlias:
+    Type: AWS::KMS::Alias
+    Properties:
+      # This will need to be synced with the `packer-rstudio` project
+      AliasName: !Sub 'alias/service-catalog/synapse/cred/${LinuxInstance}'
+      TargetKeyId: !Ref SSMParameterEncryptionKey
+
+  LinuxInstance:
+    Type: AWS::EC2::Instance
+    Metadata:
+      cfn-lint:
+        config:
+          ignore_checks:
+            - E1022
+            - W1011
+      'AWS::CloudFormation::Init':
+        configSets:
+          StartContainersJupyter:
+            - start_docker_network
+            - start_reverse_proxy
+            - make_work_dir
+            - start_jupyter_notebook
+          StartContainersRstudio:
+            - start_docker_network
+            - start_reverse_proxy
+            - make_work_dir
+            - start_rstudio_notebook
+        start_docker_network:
+          commands:
+            start_docker_network:
+              command: "docker network inspect ${NETWORK_NAME} ||  \
+                docker network create --driver bridge ${NETWORK_NAME}"
+              env:
+                NETWORK_NAME:
+                  'Fn::FindInMap': [NotebookTypeToInitParams, !Ref NotebookType, DockerNetworkName]
+        start_reverse_proxy:
+          commands:
+            start_reverse_proxy:
+              command: "docker start reverse-proxy || \
```

Comment:
`docker start..` or `docker run..`?  i'm not sure why you need `docker start`?  doesn't `docker run..` suffice?  also why run one command with `sudo` and the other without?

Reply:
> docker start.. or docker run..? i'm not sure why you need docker start? doesn't docker run.. suffice?
It's for idempotency:  If the container already exists then this command will restart it but if it doesn't then the 'docker run' command will pull down the image, create the container, and start it up.

> also why run one command with sudo and the other without?
Neither should have 'sudo'.  I will update accordingly.